File: lib/cfr/builtins/libs/dart/async/streamController.dart
import 'dart:async';
import 'dart:core';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedStreamController
    extends VMManagedBox<StreamController<dynamic>> {
  VMManagedStreamController(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['getStream'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Stream>(
            object: vmObject.stream,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getSink'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<StreamSink>(
            object: vmObject.sink, hydroState: hydroState, table: HydroTable()),
      ];
    });
    table['getIsClosed'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.isClosed,
      ];
    });
    table['getIsPaused'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.isPaused,
      ];
    });
    table['getHasListener'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hasListener,
      ];
    });
    table['add'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.add(luaCallerArguments[1]);
      return [];
    });
    table['addError'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.addError(
          maybeUnBoxAndBuildArgument<Object, dynamic>(luaCallerArguments[1],
              parentState: hydroState),
          maybeUnBoxAndBuildArgument<StackTrace?, dynamic>(
              luaCallerArguments[2],
              parentState: hydroState));
      return [];
    });
    table['close'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Future>(
            object: vmObject.close(),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getDone'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Future>(
            object: vmObject.done, hydroState: hydroState, table: HydroTable()),
      ];
    });
    table['addStream'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Future>(
            object: vmObject.addStream(
                maybeUnBoxAndBuildArgument<Stream<dynamic>, dynamic>(
                    luaCallerArguments[1],
                    parentState: hydroState),
                cancelOnError: luaCallerArguments.length >= 3
                    ? luaCallerArguments[2]['cancelOnError']
                    : null),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final StreamController vmObject;
}

void loadStreamController(
    {required HydroState hydroState, required HydroTable table}) {
  table['streamController'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    Closure? unpackedonCancel = luaCallerArguments.length >= 2
        ? luaCallerArguments[1]['onCancel']
        : null;
    Closure? unpackedonListen = luaCallerArguments.length >= 2
        ? luaCallerArguments[1]['onListen']
        : null;
    Closure? unpackedonPause = luaCallerArguments.length >= 2
        ? luaCallerArguments[1]['onPause']
        : null;
    Closure? unpackedonResume = luaCallerArguments.length >= 2
        ? luaCallerArguments[1]['onResume']
        : null;

    return [
      maybeBoxObject<StreamController>(
          object: StreamController(
              onCancel: unpackedonCancel != null
                  ? () => maybeUnBoxAndBuildArgument<FutureOr<void>, void>(
                      unpackedonCancel.dispatch(
                        [
                          luaCallerArguments[0],
                        ],
                        parentState: hydroState,
                      )[0],
                      parentState: hydroState)
                  : null,
              onListen: unpackedonListen != null
                  ? () => unpackedonListen.dispatch(
                        [
                          luaCallerArguments[0],
                        ],
                        parentState: hydroState,
                      )
                  : null,
              onPause: unpackedonPause != null
                  ? () => unpackedonPause.dispatch(
                        [
                          luaCallerArguments[0],
                        ],
                        parentState: hydroState,
                      )
                  : null,
              onResume: unpackedonResume != null
                  ? () => unpackedonResume.dispatch(
                        [
                          luaCallerArguments[0],
                        ],
                        parentState: hydroState,
                      )
                  : null,
              sync: luaCallerArguments.length >= 2
                  ? luaCallerArguments[1]['sync']
                  : null),
          hydroState: hydroState,
          table: luaCallerArguments[0])
    ];
  });
  table['streamControllerBroadcast'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    Closure? unpackedonCancel = luaCallerArguments.length >= 2
        ? luaCallerArguments[1]['onCancel']
        : null;
    Closure? unpackedonListen = luaCallerArguments.length >= 2
        ? luaCallerArguments[1]['onListen']
        : null;
    return [
      maybeBoxObject<StreamController>(
          object: StreamController.broadcast(
              onCancel: unpackedonCancel != null
                  ? () => unpackedonCancel.dispatch(
                        [
                          luaCallerArguments[0],
                        ],
                        parentState: hydroState,
                      )
                  : null,
              onListen: unpackedonListen != null
                  ? () => unpackedonListen.dispatch(
                        [
                          luaCallerArguments[0],
                        ],
                        parentState: hydroState,
                      )
                  : null,
              sync: luaCallerArguments.length >= 2
                  ? luaCallerArguments[1]['sync']
                  : null),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<StreamController>(boxer: (
      {required StreamController vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedStreamController(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/dart/async/streamController.ts
import { ISink } from "../core/sink";
import { IStackTrace } from "../core/stackTrace";
import { IEventSink } from "./eventSink";
import { IFuture } from "./future";
import { IFutureOr } from "./futureOr";
import { IStream } from "./stream";
import { IStreamConsumer } from "./streamConsumer";
import { IStreamSink } from "./streamSink";
declare const dart: {
    async: {
        streamController: <T>(
            this: void,
            streamController: IStreamController<T>,
            props: {
                onCancel?: () => IFutureOr<void> | undefined;
                onListen?: () => void | undefined;
                onPause?: () => void | undefined;
                onResume?: () => void | undefined;
                sync: boolean;
            }
        ) => IStreamController<T>;
        streamControllerBroadcast: <T>(props: {
            onCancel?: () => void | undefined;
            onListen?: () => void | undefined;
            sync: boolean;
        }) => IStreamController<T>;
    };
};
export interface IStreamController<T> {
    getStream: () => IStream<T>;
    getSink: () => IStreamSink<T>;
    getIsClosed: () => boolean;
    getIsPaused: () => boolean;
    getHasListener: () => boolean;
    add: (event: T) => void;
    addError: (error: Object, stackTrace?: IStackTrace | undefined) => void;
    close: () => IFuture<any>;
    getDone: () => IFuture<any>;
    addStream: (
        source: IStream<T>,
        props?: { cancelOnError?: boolean | undefined }
    ) => IFuture<any>;
    toString: () => string;
    getHashCode: () => number;
}
export class StreamController<T>
    implements IStreamSink<T>, IEventSink<T>, IStreamConsumer<T>, ISink<T>
{
    public constructor(props: {
        onCancel?: () => IFutureOr<void> | undefined;
        onListen?: () => void | undefined;
        onPause?: () => void | undefined;
        onResume?: () => void | undefined;
        sync?: boolean;
    }) {
        dart.async.streamController(this, {
            ...streamControllerDefaultProps,
            ...props,
        });
    }
    public static broadcast<T>(props: {
        onCancel?: () => void | undefined;
        onListen?: () => void | undefined;
        sync?: boolean;
    }): IStreamController<T> {
        return dart.async.streamControllerBroadcast({
            ...broadcastDefaultProps,
            ...props,
        });
    }
    private readonly _dart_getStream: () => IStream<T> = undefined as any;
    private readonly _dart_getSink: () => IStreamSink<T> = undefined as any;
    private readonly _dart_getIsClosed: () => boolean = undefined as any;
    private readonly _dart_getIsPaused: () => boolean = undefined as any;
    private readonly _dart_getHasListener: () => boolean = undefined as any;
    private readonly _dart_add: (event: T) => void = undefined as any;
    private readonly _dart_addError: (
        error: Object,
        stackTrace?: IStackTrace | undefined
    ) => void = undefined as any;
    private readonly _dart_close: () => IFuture<any> = undefined as any;
    private readonly _dart_getDone: () => IFuture<any> = undefined as any;
    private readonly _dart_addStream: (
        source: IStream<T>,
        props?: { cancelOnError?: boolean | undefined }
    ) => IFuture<any> = undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    public getStream(): IStream<T> {
        return this._dart_getStream();
    }
    public getSink(): IStreamSink<T> {
        return this._dart_getSink();
    }
    public getIsClosed(): boolean {
        return this._dart_getIsClosed();
    }
    public getIsPaused(): boolean {
        return this._dart_getIsPaused();
    }
    public getHasListener(): boolean {
        return this._dart_getHasListener();
    }
    public add(event: T): void {
        return this._dart_add(event);
    }
    public addError(error: Object, stackTrace?: IStackTrace | undefined): void {
        return this._dart_addError(error, stackTrace);
    }
    public close(): IFuture<any> {
        return this._dart_close();
    }
    public getDone(): IFuture<any> {
        return this._dart_getDone();
    }
    public addStream(
        source: IStream<T>,
        props?: { cancelOnError?: boolean | undefined }
    ): IFuture<any> {
        return this._dart_addStream(source, props);
    }
    public toString(): string {
        return this._dart_toString();
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
}
const streamControllerDefaultProps = {
    sync: false,
};
const broadcastDefaultProps = {
    sync: false,
};
